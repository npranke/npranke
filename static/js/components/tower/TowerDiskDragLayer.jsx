import { DragLayer } from 'react-dnd'
import PropTypes from 'prop-types'

export function collect(monitor) {
    return {
        sourceClientOffset: monitor.getSourceClientOffset(),
    }
}

export function TowerDiskDragLayer(props) {
    const previewStyle = props.sourceClientOffset ? {
        height: props.height,
        width: props.width,
        transform: (
            'translate('
            + `${props.sourceClientOffset.x}px, `
            + `${props.sourceClientOffset.y}px`
            + ')'
        ),
    } : { display: 'none' }

    return (
        <div
            className="disk-drag-layer"
            data-testid="disk-drag-layer-elem"
            style={ {
                pointerEvents: 'none',
                position: 'fixed',
                zIndex: 10,
                left: 0,
                top: 0,
                width: '100%',
                height: '100%',
            } }
        >
            <div
                className="disk dragging-preview"
                data-testid="disk-drag-preview-elem"
                style={ previewStyle }
            >
                { props.image }
            </div>
        </div>
    )
}

TowerDiskDragLayer.defaultProps = {
    height: 0,
    sourceClientOffset: null,
    width: 0,
}

TowerDiskDragLayer.propTypes = {
    height: PropTypes.number,
    image: PropTypes.node.isRequired,
    sourceClientOffset: PropTypes.shape({
        x: PropTypes.number,
        y: PropTypes.number,
    }),
    width: PropTypes.number,
}

export default DragLayer(collect)(TowerDiskDragLayer)
