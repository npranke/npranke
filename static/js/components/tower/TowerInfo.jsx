import React from 'react'

import { getSendEventHandler } from '@utils'

function TowerInfo() {
    return (
        <div className="tower-info">
            <p>
                Tower is an implementation of the Tower of Hanoi puzzle.
            </p>
            <p>
                Drag and drop disks to move the tower from origin to target.
                Larger disks can&#8217;t be placed on top of smaller disks.
                Try to move the tower in the fewest moves possible. The
                minimum number of moves is&ensp;
                <span className="math">
                    2<sup><i>n</i></sup> - 1
                </span>, where&ensp;
                <span className="math">
                    <i>n</i>
                </span>&ensp;is the number of disks in the tower.
            </p>
            <p>
                Tower is built with images of Jupiter from NASA&#8217;s Solar
                System Exploration Program. Each tower picture was taken with
                the JunoCam on the Juno spacecraft. Learn more about Jupiter,
                Juno, and other solar system missions by visiting&nbsp;
                <a
                    onClick={ getSendEventHandler(
                        'tower',
                        'navigate',
                        'NASA Solar System Exploration Program - Jupiter',
                    ) }
                    href={
                        'https://solarsystem.nasa.gov/planets/'
                        + 'jupiter/overview/'
                    }
                    target="_blank"
                    rel="noreferrer noopener external"
                >
                    NASA&#8217;s Solar System Exploration Program
                </a>.
            </p>
            <p>
                The images used in Tower are courtesy of
                NASA/JPL-Caltech. Discover more images by visiting
                the&nbsp;
                <a
                    onClick={ getSendEventHandler(
                        'tower',
                        'navigate',
                        'JPL Image Gallery',
                    ) }
                    href="https://www.jpl.nasa.gov/images"
                    target="_blank"
                    rel="noreferrer noopener external"
                >
                    JPL Image Gallery
                </a>.
            </p>
        </div>
    )
}

export default TowerInfo
