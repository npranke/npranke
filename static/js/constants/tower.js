import HTML5Backend from 'react-dnd-html5-backend'
import { MouseTransition, TouchTransition } from 'react-dnd-multi-backend'
import TouchBackend from 'react-dnd-touch-backend'

import EmptyDiscPreview from '@img/tower-empty-disc-preview.gif'

import Landscape1of2 from '@img/tower-landscape-22931-two-1-of-2.jpg'
import Landscape2of2 from '@img/tower-landscape-22931-two-2-of-2.jpg'
import Portrait1of2 from '@img/tower-portrait-22931-two-1-of-2.jpg'
import Portrait2of2 from '@img/tower-portrait-22931-two-2-of-2.jpg'

import Landscape1of3 from '@img/tower-landscape-22938-three-1-of-3.jpg'
import Landscape2of3 from '@img/tower-landscape-22938-three-2-of-3.jpg'
import Landscape3of3 from '@img/tower-landscape-22938-three-3-of-3.jpg'
import Portrait1of3 from '@img/tower-portrait-22938-three-1-of-3.jpg'
import Portrait2of3 from '@img/tower-portrait-22938-three-2-of-3.jpg'
import Portrait3of3 from '@img/tower-portrait-22938-three-3-of-3.jpg'

import Landscape1of4 from '@img/tower-landscape-22944-four-1-of-4.jpg'
import Landscape2of4 from '@img/tower-landscape-22944-four-2-of-4.jpg'
import Landscape3of4 from '@img/tower-landscape-22944-four-3-of-4.jpg'
import Landscape4of4 from '@img/tower-landscape-22944-four-4-of-4.jpg'
import Portrait1of4 from '@img/tower-portrait-22944-four-1-of-4.jpg'
import Portrait2of4 from '@img/tower-portrait-22944-four-2-of-4.jpg'
import Portrait3of4 from '@img/tower-portrait-22944-four-3-of-4.jpg'
import Portrait4of4 from '@img/tower-portrait-22944-four-4-of-4.jpg'

import Landscape1of5 from '@img/tower-landscape-22423-five-1-of-5.jpg'
import Landscape2of5 from '@img/tower-landscape-22423-five-2-of-5.jpg'
import Landscape3of5 from '@img/tower-landscape-22423-five-3-of-5.jpg'
import Landscape4of5 from '@img/tower-landscape-22423-five-4-of-5.jpg'
import Landscape5of5 from '@img/tower-landscape-22423-five-5-of-5.jpg'
import Portrait1of5 from '@img/tower-portrait-22423-five-1-of-5.jpg'
import Portrait2of5 from '@img/tower-portrait-22423-five-2-of-5.jpg'
import Portrait3of5 from '@img/tower-portrait-22423-five-3-of-5.jpg'
import Portrait4of5 from '@img/tower-portrait-22423-five-4-of-5.jpg'
import Portrait5of5 from '@img/tower-portrait-22423-five-5-of-5.jpg'

export const DRAG_AND_DROP_BACKEND_PIPELINE = {
    backends: [
        {
            backend: HTML5Backend,
        },
        {
            backend: TouchBackend,
            preview: false,
            transition: TouchTransition,
        },
        {
            backend: HTML5Backend,
            preview: false,
            transition: MouseTransition,
        },
    ],
}

export const EMPTY_PREVIEW_SOURCE = EmptyDiscPreview

export const LAYOUTS = ['landscape', 'portrait']

export const LOCATIONS = ['origin', 'buffer', 'target']

export const TOWER_SOURCES = {
    2: {
        disks: {
            landscape: [
                Landscape1of2,
                Landscape2of2,
            ],
            portrait: [
                Portrait1of2,
                Portrait2of2,
            ],
        },
        sourceid: 22931,
    },
    3: {
        disks: {
            landscape: [
                Landscape1of3,
                Landscape2of3,
                Landscape3of3,
            ],
            portrait: [
                Portrait1of3,
                Portrait2of3,
                Portrait3of3,
            ],
        },
        sourceid: 22938,
    },
    4: {
        disks: {
            landscape: [
                Landscape1of4,
                Landscape2of4,
                Landscape3of4,
                Landscape4of4,
            ],
            portrait: [
                Portrait1of4,
                Portrait2of4,
                Portrait3of4,
                Portrait4of4,
            ],
        },
        sourceid: 22944,
    },
    5: {
        disks: {
            landscape: [
                Landscape1of5,
                Landscape2of5,
                Landscape3of5,
                Landscape4of5,
                Landscape5of5,
            ],
            portrait: [
                Portrait1of5,
                Portrait2of5,
                Portrait3of5,
                Portrait4of5,
                Portrait5of5,
            ],
        },
        sourceid: 22423,
    },
}
