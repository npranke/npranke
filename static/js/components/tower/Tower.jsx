import { DragDropContext } from 'react-dnd'
import HTML5toTouch from 'react-dnd-multi-backend/lib/HTML5toTouch'
import MultiBackend from 'react-dnd-multi-backend'
import PropTypes from 'prop-types'
import React from 'react'

import TowerLocation from '@components/tower/TowerLocation'
import TowerSettings from '@components/tower/TowerSettings'
import withPortraitListener from '@components/hoc/PortraitListener'

import { LAYOUTS, LOCATIONS, TOWER_SOURCES } from '@constants/tower'

import { getSendEventHandler, sendEvent } from '@utils'

export class Tower extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            isComplete: false,
            isTimeRunning: false,
            shouldTimeReset: false,
            moves: 0,
            disks: 3,
            origin: [],
            buffer: [],
            target: [],
        }

        this.moveDisk = this.moveDisk.bind(this)
        this.updateDisks = this.updateDisks.bind(this)

        this.towers = Object.keys(TOWER_SOURCES).reduce(
            (towerAccumulator, tower) => {
                towerAccumulator[tower] = {
                    clickHandler: getSendEventHandler(
                        'tower',
                        'navigate',
                        `PIA${TOWER_SOURCES[tower].sourceid}`,
                    ),
                    link: (
                        'https://www.jpl.nasa.gov/spaceimages/details.php?'
                        + `id=PIA${TOWER_SOURCES[tower].sourceid}`
                    ),
                }

                LAYOUTS.forEach((layout) => {
                    towerAccumulator[tower][layout] = TOWER_SOURCES[
                        tower
                    ].disks[layout].map((disksource, index) => {
                        const description = (
                            'NASA/JPL Juno PIA'
                            + `${TOWER_SOURCES[tower].sourceid}, `
                            + `disk ${index + 1} of ${tower}`
                        )

                        return {
                            diskid: `${index + 1}`,
                            image: (
                                <img
                                    src={ disksource }
                                    className="disk-image"
                                    alt={ description }
                                    title={ description }
                                />
                            ),
                        }
                    })
                })

                return towerAccumulator
            },
            {},
        )
    }

    componentDidMount() {
        this.originTimeout = setTimeout(() => {
            this.setState((state) => {
                return {
                    origin: Array.from(
                        { length: state.disks },
                        (value, index) => { return `${index + 1}` },
                    ).reverse(),
                }
            })
        }, 400)
    }

    componentWillUnmount() {
        clearTimeout(this.originTimeout)
    }

    moveDisk(diskid, fromLocation, toLocation) {
        this.setState((state) => {
            const fromLocationDisks = state[fromLocation].filter(
                (fromDisk) => {
                    return fromDisk !== diskid
                },
            )
            const toLocationDisks = state[toLocation].concat([diskid])
            const isComplete = (
                toLocation === 'target'
                && toLocationDisks.length === state.disks
            )
            const isTimeRunning = !isComplete

            sendEvent('tower', 'move', 'disk')

            if (isComplete) {
                sendEvent(
                    'tower',
                    'complete',
                    `disks-${state.disks}`,
                    state.moves + 1,
                )
            }

            return {
                isComplete,
                isTimeRunning,
                shouldTimeReset: false,
                moves: state.moves + 1,
                [fromLocation]: fromLocationDisks,
                [toLocation]: toLocationDisks,
            }
        })
    }

    updateDisks(disks) {
        const updatedDisks = parseInt(disks, 10)

        sendEvent('tower', 'select', `disks-${updatedDisks}`)

        this.setState(() => {
            this.originTimeout = setTimeout(() => {
                this.setState({
                    origin: Array.from(
                        { length: updatedDisks },
                        (value, index) => { return `${index + 1}` },
                    ).reverse(),
                })
            }, 200)

            return {
                isComplete: false,
                isTimeRunning: false,
                shouldTimeReset: true,
                moves: 0,
                disks: updatedDisks,
                origin: [],
                buffer: [],
                target: [],
            }
        })
    }

    render() {
        const towerLocations = LOCATIONS.map((location) => {
            return (
                <React.Fragment key={ location }>
                    <div className="location-wrapper">
                        <TowerLocation
                            location={ location }
                            diskids={ this.state[location] }
                            disks={ this.state.disks }
                            tower={ this.towers[this.state.disks] }
                            moveDisk={ this.moveDisk }
                            isComplete={ this.state.isComplete }
                            isPortrait={ this.props.isPortrait }
                        />
                    </div>
                </React.Fragment>
            )
        })

        return (
            <div className="tower">
                <div className="settings-wrapper">
                    <TowerSettings
                        disks={ this.state.disks }
                        isTimeRunning={ this.state.isTimeRunning }
                        shouldTimeReset={ this.state.shouldTimeReset }
                        moves={ this.state.moves }
                        updateDisks={ this.updateDisks }
                    />
                </div>
                <div className="locations-wrapper">
                    <div className="locations-container">
                        { towerLocations }
                    </div>
                </div>
            </div>
        )
    }
}

Tower.defaultProps = {
    isPortrait: false,
}

Tower.propTypes = {
    isPortrait: PropTypes.bool,
}

export default DragDropContext(MultiBackend(HTML5toTouch))(
    withPortraitListener(Tower),
)
