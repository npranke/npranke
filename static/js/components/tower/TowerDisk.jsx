import { DragPreviewImage, DragSource } from 'react-dnd'
import PropTypes from 'prop-types'
import React from 'react'

import { EMPTY_PREVIEW_SOURCE } from '@constants/tower'

export const spec = {
    beginDrag: (props) => {
        return { diskid: props.diskid }
    },
    endDrag: (props, monitor) => {
        if (monitor.didDrop()) {
            const result = monitor.getDropResult()

            props.moveDisk(props.diskid, props.location, result.toLocation)
        }
    },
    canDrag: (props) => {
        return (
            !props.isComplete
            && props.diskids.indexOf(props.diskid) === props.diskids.length - 1
        )
    },
}

export function collect(connect, monitor) {
    return {
        connectDragPreview: connect.dragPreview(),
        connectDragSource: connect.dragSource(),
        canDrag: monitor.canDrag(),
        isDragging: monitor.isDragging(),
    }
}

export function TowerDisk(props) {
    const diskDragSource = props.connectDragSource(
        <div
            className={
                'disk'
                + `${props.canDrag ? ' candrag' : ''}`
                + `${props.isDragging ? ' dragging' : ''}`
            }
        >
            { props.image }
        </div>,
    )

    return (
        <React.Fragment key={ props.diskid }>
            <DragPreviewImage
                connect={ props.connectDragPreview }
                src={ EMPTY_PREVIEW_SOURCE }
            />
            { diskDragSource }
        </React.Fragment>
    )
}

TowerDisk.defaultProps = {
    canDrag: false,
    connectDragPreview: undefined,
    connectDragSource: undefined,
    diskids: [],
    height: 0,
    isComplete: false,
    isDragging: false,
    width: 0,
}

TowerDisk.propTypes = {
    canDrag: PropTypes.bool,
    connectDragPreview: PropTypes.func,
    connectDragSource: PropTypes.func,
    diskid: PropTypes.string.isRequired,
    diskids: PropTypes.arrayOf(PropTypes.string),
    height: PropTypes.number,
    image: PropTypes.node.isRequired,
    isComplete: PropTypes.bool,
    isDragging: PropTypes.bool,
    location: PropTypes.string.isRequired,
    moveDisk: PropTypes.func.isRequired,
    width: PropTypes.number,
}

export default DragSource('disk', spec, collect)(TowerDisk)
