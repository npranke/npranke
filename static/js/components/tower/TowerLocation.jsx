import { DropTarget } from 'react-dnd'
import PropTypes from 'prop-types'
import React from 'react'

import TowerDisk from '@components/tower/TowerDisk'

import useOffsetListener from '@components/hooks/OffsetListener'

export const spec = {
    drop: (props) => {
        return { toLocation: props.location }
    },
    canDrop: (props, monitor) => {
        const diskid = parseInt(monitor.getItem().diskid, 10)
        const topDisk = parseInt(props.diskids[props.diskids.length - 1], 10)

        return (
            !props.isComplete
            && (props.diskids.length === 0 || diskid < topDisk)
        )
    },
}

export function collect(connect) {
    return {
        connectDropTarget: connect.dropTarget(),
    }
}

export function TowerLocation(props) {
    const locationRef = React.useRef(null)
    const { height, width } = useOffsetListener(locationRef)

    let diskHeight, diskWidth, scaleFactor

    if (props.isPortrait) {
        diskHeight = 904
        diskWidth = 329
        scaleFactor = 1

        if (diskHeight > height) {
            scaleFactor = height / diskHeight
            diskWidth *= scaleFactor
            diskHeight = height
        }

        const availableWidth = width * 0.7
        const disksWidth = diskWidth * props.disks

        if (disksWidth > availableWidth) {
            scaleFactor = availableWidth / disksWidth
            diskWidth = availableWidth / props.disks
            diskHeight *= scaleFactor
        }
    } else {
        diskHeight = 329
        diskWidth = 904
        scaleFactor = 1

        if (diskWidth > width) {
            scaleFactor = width / diskWidth
            diskHeight *= scaleFactor
            diskWidth = width
        }

        const availableHeight = height * 0.7
        const disksHeight = diskHeight * props.disks

        if (disksHeight > availableHeight) {
            scaleFactor = availableHeight / disksHeight
            diskHeight = availableHeight / props.disks
            diskWidth *= scaleFactor
        }
    }

    let towerDisks = props.tower[
        props.isPortrait ? 'portrait' : 'landscape'
    ].filter((disk) => {
        return props.diskids.indexOf(disk.diskid) !== -1
    }).map((disk) => {
        return (
            <React.Fragment key={ disk.diskid }>
                <div
                    className="disk-wrapper"
                    style={ { height: diskHeight, width: diskWidth } }
                >
                    <TowerDisk
                        diskid={ disk.diskid }
                        diskids={ props.diskids }
                        image={ disk.image }
                        isComplete={ props.isComplete }
                        moveDisk={ props.moveDisk }
                        location={ props.location }
                    />
                </div>
            </React.Fragment>
        )
    })

    if (props.isPortrait) {
        towerDisks.reverse()
    }

    if (props.isComplete) {
        towerDisks = (
            <a
                className="location-section"
                onClick={ props.tower.clickHandler }
                href={ props.tower.link }
                target="_blank"
                rel="noreferrer noopener external"
                aria-label="NASA/JPL Space Images"
            >
                { towerDisks }
            </a>
        )
    }

    towerDisks = (
        <div
            className={
                'location-section disks '
                + `${props.location} `
                + `${props.isComplete ? 'complete' : 'notcomplete'}`
            }
        >
            { towerDisks }
        </div>
    )

    return props.isPortrait ? props.connectDropTarget(
        <div className="location" ref={ locationRef }>
            <div className="location-section name">
                <div className="name-text">
                    { props.location }
                </div>
            </div>
            { towerDisks }
        </div>,
    ) : props.connectDropTarget(
        <div className="location" ref={ locationRef }>
            { towerDisks }
            <div className="location-section name">
                <div className="name-text">
                    { props.location }
                </div>
            </div>
        </div>,
    )
}

TowerLocation.defaultProps = {
    connectDropTarget: undefined,
    diskids: [],
    isComplete: false,
    isPortrait: false,
}

TowerLocation.propTypes = {
    connectDropTarget: PropTypes.func,
    diskids: PropTypes.arrayOf(PropTypes.string),
    disks: PropTypes.number.isRequired,
    isComplete: PropTypes.bool,
    isPortrait: PropTypes.bool,
    location: PropTypes.string.isRequired,
    moveDisk: PropTypes.func.isRequired,
    tower: PropTypes.shape({
        clickHandler: PropTypes.func.isRequired,
        landscape: PropTypes.arrayOf(
            PropTypes.shape({
                diskid: PropTypes.string.isRequired,
                image: PropTypes.node.isRequired,
            }).isRequired,
        ).isRequired,
        link: PropTypes.string.isRequired,
        portrait: PropTypes.arrayOf(
            PropTypes.shape({
                diskid: PropTypes.string.isRequired,
                image: PropTypes.node.isRequired,
            }).isRequired,
        ).isRequired,
    }).isRequired,
}

export default DropTarget('disk', spec, collect)(TowerLocation)
