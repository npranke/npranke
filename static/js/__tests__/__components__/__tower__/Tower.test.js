import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import React from 'react'

import { Tower } from '@components/tower/Tower'

import * as utils from '@utils'

Enzyme.configure({ adapter: new Adapter() })

describe('Tower', () => {
    test('has settings section', () => {
        const tower = shallow(<Tower />)

        expect(
            tower.find(
                '.tower .settings-wrapper TowerSettings',
            ).exists(),
        ).toBe(true)
    })

    test('has one locations container', () => {
        const tower = shallow(<Tower />)

        expect(
            tower.find(
                '.tower .locations-wrapper .locations-container',
            ).length,
        ).toBe(1)
    })

    test('has three location sections', () => {
        const tower = shallow(<Tower />)

        expect(
            tower.find(
                '.tower .location-wrapper TowerLocation',
            ).length,
        ).toBe(3)
    })

    describe('towers', () => {
        test('has four towers', () => {
            const tower = shallow(<Tower />)

            expect(
                Object.keys(tower.instance().towers),
            ).toHaveLength(4)
        })

        test('each tower has link', () => {
            const tower = shallow(<Tower />)
            const { towers } = tower.instance()

            Object.keys(towers).forEach((towerKey) => {
                expect(
                    towers[towerKey],
                ).toHaveProperty('link', expect.any(String))
            })
        })

        test('each tower has clickHandler', () => {
            const tower = shallow(<Tower />)
            const { towers } = tower.instance()

            Object.keys(towers).forEach((towerKey) => {
                expect(
                    towers[towerKey],
                ).toHaveProperty('clickHandler', expect.any(Function))
            })
        })

        test('each tower has landscape disks', () => {
            const tower = shallow(<Tower />)
            const { towers } = tower.instance()

            Object.keys(towers).forEach((towerKey) => {
                expect(
                    towers[towerKey],
                ).toHaveProperty('landscape', expect.any(Array))
                expect(
                    towers[towerKey].landscape.length,
                ).toBe(parseInt(towerKey, 10))

                towers[towerKey].landscape.forEach((disk) => {
                    expect(
                        parseInt(disk.diskid, 10),
                    ).toBeLessThan(parseInt(towerKey, 10) + 1)
                    expect(
                        disk.image,
                    ).toBeTruthy()
                })
            })
        })

        test('each tower has portrait disks', () => {
            const tower = shallow(<Tower />)
            const { towers } = tower.instance()

            Object.keys(towers).forEach((towerKey) => {
                expect(
                    towers[towerKey],
                ).toHaveProperty('portrait', expect.any(Array))
                expect(
                    towers[towerKey].landscape.length,
                ).toBe(parseInt(towerKey, 10))

                towers[towerKey].landscape.forEach((disk) => {
                    expect(
                        parseInt(disk.diskid, 10),
                    ).toBeLessThan(parseInt(towerKey, 10) + 1)
                    expect(
                        disk.image,
                    ).toBeTruthy()
                })
            })
        })
    })

    describe('originTimeout in component lifecycle', () => {
        beforeAll(() => {
            jest.useFakeTimers()

            window.clearTimeout = jest.spyOn(window, 'clearTimeout')
        })

        afterEach(() => {
            jest.clearAllTimers()

            window.clearTimeout.mockClear()
        })

        afterAll(() => {
            window.clearTimeout.mockRestore()
        })

        test('originTimeout is defined when component did mount', () => {
            const tower = shallow(<Tower />, { disableLifecycleMethods: true })

            expect(
                tower.instance().originTimeout,
            ).not.toBeDefined()

            tower.instance().componentDidMount()

            expect(
                tower.instance().originTimeout,
            ).toBeDefined()
        })

        test('originTimeout sets origin when component did mount', () => {
            const tower = shallow(<Tower />, { disableLifecycleMethods: true })

            expect(
                tower.state('origin'),
            ).toEqual([])

            tower.instance().componentDidMount()

            jest.advanceTimersByTime(400)

            expect(
                tower.state('origin'),
            ).toEqual(['3', '2', '1'])
        })

        test('originTimeout is cleared when component will unmount', () => {
            const tower = shallow(<Tower />)

            tower.instance().componentWillUnmount()

            expect(window.clearTimeout).toHaveBeenCalledTimes(1)
        })
    })

    describe('updateDisks()', () => {
        test('sets updated disks', () => {
            const tower = shallow(<Tower />)

            tower.instance().updateDisks('5')

            expect(
                tower.state('disks'),
            ).toBe(5)
        })

        test('sets originTimeout', () => {
            const tower = shallow(<Tower />)

            tower.instance().updateDisks('2')

            expect(
                tower.instance().originTimeout,
            ).toBeDefined()
        })

        test('originTimeout resets origin with updated diskids', () => {
            jest.useFakeTimers()

            const tower = shallow(<Tower />)

            tower.instance().updateDisks('4')

            jest.advanceTimersByTime(200)

            expect(
                tower.state('origin'),
            ).toEqual(['4', '3', '2', '1'])
        })
    })

    describe('moveDisk()', () => {
        test('moves disk when not completing tower', () => {
            const tower = shallow(<Tower />)

            tower.setState({
                isTimeRunning: true,
                moves: 2,
                disks: 3,
                origin: ['3'],
                buffer: ['2'],
                target: ['1'],
            })

            tower.instance().moveDisk('1', 'target', 'buffer')

            expect(tower.state()).toMatchObject({
                isComplete: false,
                isTimeRunning: true,
                moves: 3,
                origin: ['3'],
                buffer: ['2', '1'],
                target: [],
            })
        })

        test('moves disk when completing tower', () => {
            const tower = shallow(<Tower />)

            tower.setState({
                isTimeRunning: true,
                moves: 6,
                disks: 3,
                origin: ['1'],
                buffer: [],
                target: ['3', '2'],
            })

            tower.instance().moveDisk('1', 'origin', 'target')

            expect(tower.state()).toMatchObject({
                isComplete: true,
                isTimeRunning: false,
                moves: 7,
                origin: [],
                buffer: [],
                target: ['3', '2', '1'],
            })
        })
    })

    describe('sending tower events', () => {
        beforeAll(() => {
            utils.getSendEventHandler = jest.spyOn(
                utils,
                'getSendEventHandler',
            )
            utils.sendEvent = jest.spyOn(utils, 'sendEvent')
        })

        afterEach(() => {
            utils.getSendEventHandler.mockClear()
            utils.sendEvent.mockClear()
        })

        afterAll(() => {
            utils.getSendEventHandler.mockRestore()
            utils.sendEvent.mockRestore()
        })

        test('when disks is updated', () => {
            const tower = shallow(<Tower />)

            tower.instance().updateDisks('2')

            expect(
                utils.sendEvent,
            ).toHaveBeenNthCalledWith(
                1,
                'tower',
                'select',
                'disks-2',
            )
        })

        test('when disk is moved', () => {
            const tower = shallow(<Tower />)

            tower.setState({
                disks: 4,
                origin: ['4', '3', '2', '1'],
                buffer: [],
                target: [],
            })

            tower.instance().moveDisk('1', 'origin', 'buffer')

            expect(
                utils.sendEvent,
            ).toHaveBeenNthCalledWith(
                1,
                'tower',
                'move',
                'disk',
            )
        })

        test('when tower is completed', () => {
            const tower = shallow(<Tower />)

            tower.setState({
                isTimeRunning: true,
                moves: 6,
                disks: 3,
                origin: ['1'],
                buffer: [],
                target: ['3', '2'],
            })

            tower.instance().moveDisk('1', 'origin', 'target')

            expect(
                utils.sendEvent,
            ).toHaveBeenNthCalledWith(
                2,
                'tower',
                'complete',
                'disks-3',
                7,
            )
        })

        test('with getSendEventHandler for completed tower', () => {
            shallow(<Tower />)

            expect(
                utils.getSendEventHandler,
            ).toHaveBeenCalledTimes(4)
            expect(
                utils.getSendEventHandler,
            ).toHaveBeenNthCalledWith(
                1,
                'tower',
                'navigate',
                'PIA22931',
            )
        })
    })
})

describe('Tower snapshot', () => {
    test('matches snapshot', () => {
        const tower = shallow(<Tower />)

        expect(tower).toMatchSnapshot()
    })
})
