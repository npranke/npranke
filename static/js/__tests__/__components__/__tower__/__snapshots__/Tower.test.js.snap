// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tower snapshot matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="tower"
  >
    <div
      class="settings-wrapper"
    >
      <div
        class="settings"
      >
        <div
          class="settings-inner"
        >
          <div
            class="left-text"
          >
            <span
              class="disks-text"
            >
              <span
                class="disks-portion"
                id="disks-label"
              >
                disks: 
              </span>
              <span
                aria-labelledby="disks-label"
                class="disks-portion"
                id="disks-radiogroup"
                role="radiogroup"
              >
                <input
                  aria-checked="false"
                  id="two-disks"
                  name="disks"
                  type="radio"
                  value="2"
                />
                <label
                  class="disks-portion number"
                  data-value="2"
                  for="two-disks"
                  id="two-disks-label"
                  tabindex="-1"
                >
                  2
                </label>
                <input
                  aria-checked="true"
                  checked=""
                  id="three-disks"
                  name="disks"
                  type="radio"
                  value="3"
                />
                <label
                  class="disks-portion number"
                  data-value="3"
                  for="three-disks"
                  id="three-disks-label"
                  tabindex="0"
                >
                  3
                </label>
                <input
                  aria-checked="false"
                  id="four-disks"
                  name="disks"
                  type="radio"
                  value="4"
                />
                <label
                  class="disks-portion number"
                  data-value="4"
                  for="four-disks"
                  id="four-disks-label"
                  tabindex="-1"
                >
                  4
                </label>
                <input
                  aria-checked="false"
                  id="five-disks"
                  name="disks"
                  type="radio"
                  value="5"
                />
                <label
                  class="disks-portion number"
                  data-value="5"
                  for="five-disks"
                  id="five-disks-label"
                  tabindex="-1"
                >
                  5
                </label>
              </span>
            </span>
          </div>
          <div
            class="right-text"
          >
            <span
              class="moves-text"
            >
              <span
                class="moves-portion"
              >
                moves: 
              </span>
              <span
                class="moves-portion number"
              >
                00
              </span>
            </span>
            <span
              class="time-text"
            >
              <span
                class="time-portion"
              >
                time: 
              </span>
              <span
                class="time-portion number"
              >
                00
              </span>
              <span
                class="time-portion"
              >
                 : 
              </span>
              <span
                class="time-portion number"
              >
                00
              </span>
              <span
                class="time-portion"
              >
                 : 
              </span>
              <span
                class="time-portion number"
              >
                00
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="locations-wrapper"
    >
      <div
        class="locations-container"
      >
        <div
          class="location-wrapper"
        >
          <div
            class="location"
          />
        </div>
        <div
          class="location-wrapper"
        >
          <div
            class="location"
          />
        </div>
        <div
          class="location-wrapper"
        >
          <div
            class="location"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
